proposals/0000-visible-forall.rst

```diff
@@ -0,0 +1,1133 @@
+
+   This is a necessity to avoid parsing conflicts, with the following
+   consequences:
+
+   1. The ``'`` symbol signifies Template Haskell name quotation rather than ``DataKinds`` promotion.
+   2. The ``*`` symbol is treated as an infix operator regardless of ``-XStarIsType``.
+   3. Built-in syntax for tuples and lists is interpreted as in terms.
+      That is, ``[a]`` is a singleton list rather than the type of a list,
+      and ``(a, b)`` is a pair rather than the type of a pair.
+
+   For the above grammar to work, make ``forall`` a keyword at the term
+   level. Not guarded by any extension (same motivation as `#193
+   <https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0193-forall-keyword.rst>`_).
+
+   In name resolution, we follow the term-level rules that prioritize data
+   constructors over type constructors, and variables over type variables.
+
+   In types-in-terms, the ``forall``-bound type variables are put in the
+   term-level namespace, so that in ``\x -> f (forall x. x)``, the ``x``
+   occurrence refers to the forall-bound type variable rather than the
+   lambda-bound variable.
+
+   Extend the T2T mapping as follows:
+
+   * The types-in-terms (such as ``a -> b``, ``a => b``, ``forall a. b``) are
+     mapped to types directly, without modification aside from recursively
+     processing subterms.
+
+   Any occurrence of types-in-terms that does not undergo the T2T mapping is
+   rejected during type checking.
+
+9. When ``ViewPatterns`` are enabled, interpret ``f (a -> b) = ...``
+   as a view pattern, otherwise as ``f ((->) a b) = ...``.
+
+10. ``case ... of x -> y -> z`` is an error. We require parentheses to
+    disambiguate:
+
+    * ``case ... of (x -> y) -> z``
+    * ``case ... of x -> (y -> z)``
+
+Secondary Change: Lists and Tuples
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+11. Introduce a new extension, ``ListTupleTypeSyntax``, on by default,
+    which enables:
+
+    * ``[]`` as the list type constructor
+    * ``()`` as the unit type
+    * ``[a]`` the syntax of a list type
+    * ``(,)`` as the pair type constructor
+    * ``(a, b)`` as the syntax of a pair type
+
+    When the extension is on, these constructs retain their Haskell 2010
+    meaning, which depends on whether we are in a type-level or term-level
+    context.
+
+    When the extension is off, all of the above are interpreted as in terms:
+
+    * ``[]`` is always an empty list
+    * ``()`` is always the unit data constructor
+    * ``[a]`` is always a singleton list (not the list type)
+    * ``(,)`` is always the pair data constructor (not the type constructor)
+    * ``(a, b)`` is always a pair (not the type of a pair)
+
+    Likewise for tuples of higher arities.
+
+    Export the following synonyms from the ``GHC.Exts`` module::
+
+      type List = []
+      type Unit = ()
+      type Tuple2 = (,)
+      type Tuple3 = (,,)
+      type Tuple4 = (,,,)
+      ... -- up to the maximum tuple arity
+
```

Comment:
I vaguely remember that those syntactic constructs are always in scope, even without a Prelude. Does this extension change that?

Reply:
With the extension, `[]` `()` `(,)` are always in scope, but refer to the data constructors regardless of type/term context.